vtpm/src/char.rs

```diff
@@ -0,0 +1,267 @@
+    TPMCharFailedWrite(#[source] anyhow::Error),
+    #[error("Failed to configure TPM Char backend")]
+    CharBackendConfigureFailed(#[source] anyhow::Error)
+}
+type Result<T> = anyhow::Result<T, TPMCharError>;
+
+#[derive(PartialEq)]
+enum ChardevState {
+    ChardevStateDisconnected,
+    ChardevStateConnecting,
+    ChardevStateConnected,
+}
+
+/// Copy data in `from` into `to`, until the shortest
+/// of the two slices.
+///
+/// Return the number of bytes written.
+fn byte_copy(from: &[u8], mut to: &mut [u8]) -> usize {
+    to.write(from).unwrap()
+}
+
+pub struct SocketCharDev {
+    state: ChardevState,
+    stream: Option<UnixStream>,
+    /// Fd sent to swtpm process for Data
+    write_msgfd: RawFd,
+    /// Control Channel
+    ctrl_fd: RawFd,
+    /// Data Channel
+    data_fd: RawFd,
+   // chr_write_lock: Arc<Mutex<usize>>,
+}
+
+impl SocketCharDev {
+    pub fn new() -> Self {
+        Self {
+            state: ChardevState::ChardevStateDisconnected,
+            stream: None,
+            write_msgfd: -1,
+            ctrl_fd: -1,
+            data_fd: -1,
+            //chr_write_lock: Arc::new(Mutex::new(0)),
+        }
+    }
+
+    pub fn connect(&mut self, socket_path: &str) -> Result<isize> {
+        self.state = ChardevState::ChardevStateConnecting;
+
+        let now = Instant::now();
+
+        // Retry connecting for a full minute
+        loop {
+            match UnixStream::connect(socket_path) {
+                Ok(s) => {
+                    let fd = s.as_raw_fd();
+                    self.ctrl_fd = fd;
+                    self.stream = Some(s);
+                    self.state = ChardevState::ChardevStateConnected;
+                    debug!("Connected to vTPM socket path : {:?}\n", socket_path);
+                    return Ok(0);
+                }
+                Err(_e) => {}
+            };
+            sleep(Duration::from_millis(100));
+
+            if now.elapsed().as_secs() >= 60 {
+                break;
+            }
+        }
+        Err(TPMCharError::TPMCharCannotConnect(anyhow!(
+            "Failed to connect to vTPM socket path"
+        )))
+    }
+
+    pub fn set_datafd(&mut self, fd: RawFd) {
+        self.data_fd = fd;
+    }
+
+    pub fn set_msgfd(&mut self, fd: RawFd) {
+        self.write_msgfd = fd;
+    }
+
+    pub fn chr_sync_read(&self, buf: &mut [u8], _len: usize) -> Result<usize> {
+        if self.state != ChardevState::ChardevStateConnected {
+            return Ok(0);
+        }
+        debug!("synchronous read from vTPM Socket");
+        let size = recv(self.ctrl_fd, buf, MsgFlags::empty()).map_err(|e| {
+            TPMCharError::TPMCharFailedRead(anyhow!(
+                "Failed to read from vTPM socket. Error Code {:?}",
+                e
+            ))
+        })?;
+        debug!("sync read completed");
+        Ok(size as usize)
+    }
+
+    pub fn send_full(&self, buf: &mut [u8], _len: usize) -> Result<usize> {
```

Comment:
Can you abstract out a `tpm` trait so that we can easily plug in other backend in the future?